docs/sstrings.5
.\" Copyright 2005 by Pablo Yanez Trujillo <pabloy@pcpool.mathematik.uni-freiburg.de
.\" The safe Strings Library Version 1.0.5
.\" 
.\" This is free software. Please read the file ../COPYING if you
.\" want to use/edit/distribuite this source file.
.\" This source file is protected by the GNU GPL-2
.\" NOTE: There is NO  warranty; not even for MERCHANTABILITY or 
.\" FITNESS FOR A PARTICULAR PURPOSE.
.TH "sstrings" "5" "July 2005" "Version 1.0.5" "Linux Programmer's Manual -- Safe Strings Library"
.SH "SINOPSYS"
The
.B Safe Strings Libraray
is a simple collection of C functions that handle with strings in a safe way.
All function were written with ANSI C (C99) functions are were tested before
this current release was made, but this doesn't mean that all function are bug free.
So, if you find erros and bugs, please tell me about them at pabloy@pcpool.mathematik.uni-freiburg.de

I use the name \fBSafe Strings\fR because the functions allocate and reallocate only the needed space, 
each time before writting, expanding or cutting the memory those functions proof that there is enough 
space to execute they jobs. So you don't need (as programmer) to take care about this, because this 
is silly work when working with C String.

This function should be used only with C. C++ has std::string and std::string should be used there, so don't use
this funtion with C++.
.SH "PROGRAMMING WITH SSTRINGS"
Before you begin porgramming with this library you should know this:
.LP
.TP
.SS "How to use this libraray"
You may use these functions as a library (shared or static) or you can add the whole source code to your projects.
The code is released under the term of the GPL-2.

.LP
.SS "Writing code and including source files"
The only thing you need to access the library, is to include \fIsstrings.h\fR into your source files. Note that you
don't need to include \fIstdio.h\fR, \fIstdlib.h\fR and \fIctype.h\fR because they are included in \fIsstrings.h\fR.

You must call \fIfree(3)\fR with each variable that is used as a pointer of any result of one of sstrings functions.
For example:

.RS
.nf
	/* sourcecode: helloworld.c */

	#include <sstrings.h>
	#include <stdio.h>

	int main(void)
	{
		char* x = "Hello";
		char y[] = "World!";
		char* helloworld = NULL;

		helloworld = appendstring(&helloworld, x); /* assume everything goes right, this is so */
		helloworld = appendstring(&helloworld, " "); /* a dirty code */
		helloworld = appendstring(&helloworld, y);

		printf("%s\\n", helloworld); /* output: Hello World */
		
		/* you don't need top free x and y, but helloworld */

		free(helloworld); /* YOU MUST DO THIS BY YOURSELF */

		return 0;
	}
.fi
.RE
.LP
.BR "Linking source code with sstrings"
Use \fI-lsstrings\fR to link this library. Example:

\&	$ gcc -ohelloworld -lsstrings helloworld.c

.SS "Online documentation"
You find an online documentation under 

http://pcpool.mathematik.uni-freiburg.de/~pabloy/projects/sstrings/
.SS "Note"
Avoid using the same variable when you use one of this functions. somefunction(&str, str, other parm.); will work in most of the 
cases but if realloc changes the adress then you will posibly get a segmentation fault!

.SH "BUGS"
I din't find any bugs at all, please send me a bug report at pabloy@pcpool.mathematik.uni-freiburg.de
if you find one. Thank you

.SH "ERROR CONSTANTS"
When the functions return NULL the variable \fIss_errno\fR is set to a error constant. In each man page of the functions
is explained which values every function may generate. Following table shows the different error constants.

.RS
.nf
 
	List of error constants

	Name              || value || Description
	=========================================
	\fBSS_NO_ERROR\fR          0        describes the state of no error
	\fBSS_NO_MEM\fR            1        cannot execute malloc because of the memory
	\fBSS_NO_EXPAND\fR         2        cannot execute realloc and expand the buffer
	\fBSS_NULL_BUFFER\fR       3        you passed a NULL buffer, when a string is required
	                              or a not NULL pointer is required
	\fBSS_NEEDLE_NOT_FOUND\fR  4        when the needle is not found (reasons: empty needle or
	                              the needle is really not found)
	\fBSS_NULL_TRIM\fR         5        *buffer in trim sstr*trim function is NULL. Cannot trim anything
	\fBSS_EMPTY_STR\fR         6        When a buffer contains an empty string and it shouldn't be empty.

.fi
.RE
.SH "FUNCTIONS"
I don't want to explain here the functions. Please see the \fBSEE ALSO\fR section, there are all functions listed, you can see
the manual page for each function.

For more information and documentation visit http://pcpool.mathematik.uni-freiburg.de~/pabloy/projects/sstrings

.SH "SEE ALSO"
.BR addchar (3)
.BR appendstring (3)
.BR appendnstring (3)
.BR stringappend (3)
.BR stringnappend (3)
.BR str2sstr (3)
.BR subsstr (3)
.BR strrstr (3)
.BR substr (3)
.BR subrstr (3)
.BR delsstr (3)
.BR cutsstr (3)
.BR cutrsstr (3)
.BR ltrim (3)
.BR rtrim (3)
.BR trim (3)
.BR sstrltrim (3)
.BR sstrrtrim (3)
.BR sstrtrim (3)
.BR insertstring (3)
.BR insertnstring (3)
.BR insertrnstring (3)
.BR replacestring (3)
.BR replacerstring (3)
.BR replacenstring (3)
.BR replacenlaststring (3)
.BR replacernstring (3)
.BR replacernlaststring (3)
.BR emptystring (3)


NEWS
Sun Mar 20 00:05:26 CET 2005
  - A new version was created, 1.0.3. I saw that the 0.0.x series were bad selected, beacause
    it makes no sense to have a version 0. So I change the version release to 1.0.x. This will be
    the third release of sstrings and that's why the project becomes 1.0.3

Fri Mar 11 19:54:04 CET 2005

  - Version 0.0.2 released
  - NEWS, TODO, README, BUGS file were rewritten
  - configure script rewritten


sstrings/sstrings.c
/**
 * Filename: sstrings/sstrings.c
 * Author: Pablo Yanez Trujillo <rex@supertux.homelinux.org>
 *
 * This is free software. Please read the file ../COPYING if you
 * want to use/edit/distribuite this source file.
 * This source file is protected by the GNU GPL-2
 *
 * NOTE: There is NO  warranty; not even for MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.
**/


/**
 * Version: 1.0.3
 * Please read the TODO, NEWS, BUGS files
 *
 * Last change: 24.03.2005 by Pablo, alias Supertux
 *
 **/

#include <sstrings.h>

short ss_errno; /* This variable must be rest to SS_NO_ERROR manually by the user */



char* addchar(char** buffer, int c)
{
	size_t len;
	char* tmp;
	
	if(!buffer)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}


	if(!*buffer)
	{
		/* allocate 2 Bytes of memory */
		tmp = malloc(!c ? 1 : 2);

		if(!tmp)
		{
			ss_errno = SS_NO_MEM;
			return NULL;
		}

		*tmp = c;
		*(tmp + (!c ? 0 : 1))=0;

		*buffer = tmp;
		
	} else {

		if(!c) return *buffer;
		
		/* read all size and allocate old size+2 new characters */
		len = strlen(*buffer) + 2;

		tmp = realloc(*buffer,len);

		if(!tmp)
		{
			ss_errno = SS_NO_EXPAND;
			return NULL;
		}

		if(tmp != *buffer)
			*buffer = tmp;
		
		*(tmp + len - 2) = c;
	        *(tmp + len - 1) = 0;
	}

	return *buffer;
}




char* appendnstring(char** buffer, const char* string, size_t len)
{
	size_t buf_len, str_len;
	char* tmp, *new_tmp;

	
	if(!buffer || !string)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	/* if len > strlen(string) then set len to strlen(string) */

	str_len = strlen(string); /* hope string is 0-terminated */

	if(len > str_len)
		len = str_len;

	/* if len == 0 then do nothing and return *buffer.
	 * string may be an empty string, in this case len is 0, too.
	 * In this case, allocate only one byte of memory and set it
	 * to 0 and return it. If string is empty then len will be 0
	 */

	if(!*buffer && !len) /* ok, *buffer is NULL and len==0 and string may be empty (if not then do as if) 
				if *buffer is NULL and string is empty (the case *buffer==NULL and !*string) is 
				also there! */
	{
		tmp = NULL;
		tmp = addchar(&tmp, 0);

		if(!tmp) return NULL;

		return *buffer = tmp;

	} else if(!len) /* len==0, *buffer is not NULL, then do nothing */
		return *buffer;

	/* ok, *string != 0 and len != 0, let's rock :) */
	
	if(!*buffer)
	{
		/* allocate len+1 bytes */
		
		tmp = malloc(len + 1);

		if(!tmp)
		{
			ss_errno = SS_NO_MEM;
			return NULL;
		}

		/* what if *string==0 ? */
		if(!*string)
		{
			**buffer = *string;
			return *buffer;
		}

		/* be sure, avoid using memcpy, strcpy, etc. Set \0 by yourself */
		memmove(tmp, string, len);
		*(tmp + len) = 0; /* set \0 by yourself */

		*buffer = tmp;
		
	} else {
		/* a little bit harder. Try to reallocate the space */

		buf_len = strlen(*buffer);


		tmp = realloc(*buffer, buf_len+len+1);


		if(!tmp)
		{
			ss_errno = SS_NO_EXPAND;
			return NULL;
		}

		if(tmp != *buffer)
		{
			/* printf("\t***DEBUG: memory changed from %x to %x**\n", *buffer, tmp); */
			*buffer = tmp;
		}

		/* go to the \0 of *buffer */
		new_tmp = tmp + buf_len;

		memmove(new_tmp, string, len); /* again, set \0 by yourself */

		new_tmp += len;

		*new_tmp = 0;
	}


	return *buffer;
}




char* appendstring(char** buffer, const char* string)
{
	return appendnstring(buffer, string, string ? strlen(string) : 0);
}




char* stringnappend(char** buffer, const char* string, size_t len)
{
	size_t buf_len, str_len;
	char* tmp, *new_tmp;

	if(!buffer || !string)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	/* same cases as appendnstring */

	str_len = strlen(string); /* hope string is 0-terminated */

	if(len > str_len)
		len = str_len;

	if(!*buffer && !len)
	{
		tmp = NULL;
                tmp = addchar(&tmp, 0);
                
		if(!tmp) return NULL;
 
		return *buffer = tmp;
	} else if(!len)
		return *buffer;
	
	if(!*buffer)
	{
		*buffer = appendnstring(buffer, string, str_len);
		return *buffer;
	}

	/* don't need at this point the else anymore */
	
	/* first of all increment the allocated space and move the old
	 * content, including the \0 */

	buf_len = strlen(*buffer);

	tmp = realloc(*buffer, buf_len + len + 1);

	if(!tmp)
	{
		ss_errno = SS_NO_EXPAND;
		return NULL;
	}

	if(tmp != *buffer)
		*buffer = tmp;

	new_tmp = tmp + len;
	memmove(new_tmp, *buffer, buf_len);

	/* set \0 by yourself. *buffer is 0-terminated, but be sure */
	*(tmp + buf_len + len) = 0;

	memmove(tmp, string, len);

	return *buffer;
}




char* stringappend(char** buffer, const char* string)
{
	return stringnappend(buffer, string, string ? strlen(string) : 0);
}




char* str2sstr(const char* string)
{
	char* tmp = NULL;

	if(!string)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	tmp = appendnstring(&tmp, string, strlen(string));

	return tmp;
}




char* subsstr(char** buffer, size_t pos, size_t len)
{
	size_t buf_len;
	char* tmp;

	if(!buffer)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	} else if(!*buffer && !len)
	{
		ss_errno = SS_NO_ERROR;
		return NULL;
	} else if(*buffer && !len)
		return NULL;

	/* if len > strlen(*buffer) ==> len = strlen(*buffer) */

	buf_len = strlen(*buffer);

	if(!buf_len) 
	{
		ss_errno = SS_EMPTY_STR;
		return NULL;
	}

	if(pos > buf_len - 1) pos = buf_len - 1; /* we can make here so
						    that *buffer is returned instead of
						    changing pos */

	if(len > buf_len - pos) len = buf_len - pos;

	tmp = *buffer + pos;

	memmove(*buffer, tmp, len);

	tmp = *buffer;
	*(tmp + len) = 0;

	/* for this function and for all other functions.
	 * if realloc fails (what it may be in a very very
	 * strange case) then return the old buffer.
	 *
	 */

	tmp = realloc(*buffer, strlen(*buffer) + 1);

	if(tmp && (tmp != *buffer))
		*buffer = tmp;

	return *buffer;
	
}




char* strrstr(const char *haystack, const char *needle)
{
	char* tmp, *tmp_old;
	size_t len;
	int counter;

	
	if(!haystack || !needle)
		return NULL;

	if(!*haystack && !*needle)
		return (char*) &haystack[0]; /* gcc generates a warning */
				     /* other compiler may not compile this */
	else if(!*haystack)
		return NULL;
	else if(!*needle) 
		return (char*) &haystack[0]; /* the same */

	tmp_old = tmp = (char*) &haystack[0]; /* the same */
	len = strlen(needle);

	counter = 0;

	while(1)
	{
		tmp = strstr(tmp, needle);

		if(!tmp)
		{
			if(!counter) return NULL;
			return tmp_old;
		}

		counter++;
		
		if(strlen(tmp)==len) 
			return tmp; /* the last strlen(needle)
					   match with needle. */
		/* in this case, strlen(tmp_new > strlen(needle) */
		tmp_old = tmp;
		tmp += len; /* point to the nex character after needle */
		
	}

	return tmp_old;	

}




char* substr(const char* haystack, const char* needle)
{
	char* tmp;
	
	if(!haystack || !needle)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*needle)
		return str2sstr(haystack);

	tmp = strstr(haystack, needle);

	if(!tmp)
	{
		ss_errno = SS_NEEDLE_NOT_FOUND;
		return NULL;
	}

	
	return str2sstr(tmp);
}




char* subrstr(const char* haystack, const char* needle)
{
	char* tmp;

	if(!haystack || !needle)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*haystack || !*needle)
		return str2sstr(haystack);

	tmp = strrstr(haystack, needle);

	if(!tmp)
	{
		ss_errno = SS_NEEDLE_NOT_FOUND;
		return NULL;
	}

	return str2sstr(tmp);
}




char* delsstr(char** buffer, size_t pos, size_t len)
{
	char* tmp, *tmp2;
	size_t str_len;
	
	if(!buffer)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*buffer) return *buffer;

	str_len = strlen(*buffer);

	if(pos > str_len - 1) pos = str_len - 1; /* we can make so
						    that *buffer is returned instead of
						    changing pos */

	if(len > str_len - pos) len = str_len - pos;

	if(!len) return *buffer;

	tmp = *buffer + pos;
	tmp2 = tmp + len;

	str_len = strlen(tmp2);
	
	if(!str_len)
	{
		/* cut the 'len' last characters. */
		tmp = *buffer;
		tmp[strlen(tmp) - len] = 0;
		return *buffer;
	}

	memmove(tmp, tmp2, str_len);

	/* set 0 */
	tmp = *buffer;

	tmp[pos + str_len] = 0;
	
	str_len = strlen(tmp);

	tmp = realloc(*buffer, str_len + 1);

	if(tmp && (tmp != *buffer))
		*buffer = tmp;

	return *buffer;
}




char* cutsstr(char** buffer, const char* needle)
{
	char* tmp;
	size_t len, pos;
	
	if(!buffer || !needle)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*needle) return *buffer;

	tmp = strstr(*buffer, needle);

	if(!tmp)
	{
		ss_errno = SS_NEEDLE_NOT_FOUND;
		return *buffer;
	}

	pos = tmp - *buffer;
	len = strlen(needle);

	tmp = delsstr(buffer, pos, len);

	if(!tmp) return NULL;

	if(tmp != *buffer)
		*buffer = tmp;

	return *buffer;
}




char* cutrsstr(char** buffer, const char* needle)
{
        char* tmp;
        size_t len, pos;

        if(!buffer || !needle)
        {
                ss_errno = SS_NULL_BUFFER;
                return NULL;
        }

        if(!*needle) return *buffer;

        tmp = strrstr(*buffer, needle);

        if(!tmp)
        {
                ss_errno = SS_NEEDLE_NOT_FOUND;
                return *buffer;
        }

        pos = tmp - *buffer;
        len = strlen(needle);

        tmp = delsstr(buffer, pos, len);

        if(!tmp) return NULL;

        if(tmp != *buffer)
                *buffer = tmp;

        return *buffer;
}




char* ltrim(char* buffer)
{
	if(!buffer)
		return NULL;

	while(isspace(*buffer))
		buffer++;

	return buffer;
}




char* rtrim(char* buffer)
{
	size_t len;
	char* tmp;
	if(!buffer)
		return NULL;

	len = strlen(buffer);

	if(!len) return buffer;

	tmp = buffer + len - 1;
	
	while(isspace(*tmp) && (tmp != buffer))
	{
		*tmp = 0;
		tmp--;
	}

	*tmp = isspace(*tmp) ? 0 : *tmp;

	return buffer;
}




char* trim(char* buffer)
{
	return rtrim(ltrim(buffer));
}




char* sstrltrim(char** buffer)
{
	char* tmp;
	size_t len;
	
	if(!buffer)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*buffer)
	{
		ss_errno = SS_NULL_TRIM;
		return NULL;
	}

	tmp = ltrim(*buffer);

	len = strlen(tmp);

	memmove(*buffer, tmp, len);

	/* set \0 by yourself */
	*(*buffer + len) = 0;

	tmp = realloc(*buffer, len + 1);

	if(tmp && (tmp != *buffer))
		*buffer = tmp;

	return *buffer;
}




char* sstrrtrim(char** buffer)
{
	char* tmp;
	size_t len;

	if(!buffer)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*buffer)
	{
		ss_errno = SS_NULL_TRIM;
		return NULL;
	}

	rtrim(*buffer);

	len = strlen(*buffer);

	tmp = realloc(*buffer, len + 1);

	if(tmp && (tmp != *buffer))
		*buffer = tmp;

	return *buffer;
}




char* sstrtrim(char** buffer)
{
	char* tmp;

	if(!buffer)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!*buffer)
	{
		ss_errno = SS_NULL_TRIM;
		return NULL;
	}

	/* important, right trim before of ltrim */

	tmp = sstrrtrim(buffer);

	if(!tmp)
	{
		/* ss_errno is set by sstrrtrim */
		return NULL;
	}

	if(tmp != *buffer)
		*buffer = tmp;

	tmp = sstrltrim(buffer);

	if(!tmp)
	{
		/* ss_errno is set by sstrrtrim */
		return NULL;
	}

	if(tmp != *buffer)
		*buffer = tmp;

	return *buffer;
}





/* Functions added in the 1.0.3 version */

static char* __replacestr(char** buffer, const char* haystack, const char* needle, size_t len, size_t search)
{
	size_t buff_len[5] = {0,0,0,0,0}; /* 0 => haystack, 1 => needle, 
					 2 => move, 3=> new len,
					 4 => new found pos */

	char* found, *move, *tmp;

	
	if(!buffer || !haystack || !needle)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!len || !*buffer)
	{
		ss_errno = SS_NO_ERROR;
		return *buffer;
	}

	buff_len[0] = strlen(haystack);
	buff_len[1] = strlen(needle);

	if(len > buff_len[1])
		len = buff_len[1];

	found = search ? strstr(*buffer, haystack) : strrstr(*buffer, haystack);

	if(!found)
	{
		ss_errno = SS_NEEDLE_NOT_FOUND;
		return NULL;
	}

	move = found + buff_len[0];

	/* set move to NULL if *move==0, because nothing
	 * must be copied. */

	if(*move)
	{
		move = str2sstr(move);

		if(!move) return NULL;
	} else move = NULL;

	buff_len[2] = move ? strlen(move) : 0;
	buff_len[4] = found - *buffer;

	buff_len[3] = buff_len[4] + len + buff_len[2];

	/* resize buffer */

	tmp = realloc(*buffer, buff_len[3] + 1);

	if(!tmp)
	{
		ss_errno = SS_NO_EXPAND;
		if(move) free(move);
		return NULL;
	}

	if(tmp != *buffer)
		*buffer = tmp;

	/* set new found (after realloc *buffer may have been changed) */

	found = *buffer + buff_len[4];

	memmove(found, needle, len);

	found = *buffer + buff_len[4] + len;

	if(move)
	{
		memmove(found, move, buff_len[2]);

		found = *buffer + buff_len[3];
	}
	
	*found = 0;

	if(move) free(move);
	
	return *buffer;
}


static char* __replacelaststr(char** buffer, const char* haystack, const char* needle, size_t len, int search)
{
        int need_len;

        if(!needle)
        {
                ss_errno = SS_NULL_BUFFER;
                return NULL;
        }

        need_len = strlen(needle);

        if(need_len < len)
                len = need_len;

        if(need_len != len)
                needle += need_len - len;

        return __replacestr(buffer, haystack, needle, len, search);
}




char* insertnstring(char** buffer, const char* src, size_t pos, size_t len)
{
	char* tmp = NULL, *move;
	size_t src_len, buf_len;
	
	if(!buffer || !src)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!len)
	{
		ss_errno = SS_NO_ERROR;
		return *buffer;
	}

	src_len = strlen(src);

	if(len > src_len) len = src_len;

	if(!*buffer)
		return appendnstring(&tmp, src, len);

	buf_len = strlen(*buffer);
	
	if(pos > buf_len) pos = buf_len + 1;

	if(pos == buf_len + 1)
		return appendnstring(buffer, src, len);

	if(!pos)
		return stringnappend(buffer, src, len);

	
	tmp = realloc(*buffer, buf_len + len + 1);

	if(!tmp)
	{
		ss_errno = SS_NO_EXPAND;
		return NULL;
	}

	if(tmp != *buffer)
		*buffer = tmp;

	tmp += pos; /* set tmp to the position where it must be inserted */
	move = tmp + len; /* set move to the position where the string must be moved */

	memmove(move, tmp, buf_len - pos + 1); /* +1 to copy \0 */

	/* 2nd block */

	memmove(tmp, src, len);

	return *buffer;
}




char* insertstring(char** buffer, const char* src, size_t pos)
{
	return insertnstring(buffer, src, pos, src ? strlen(src) : 0);
}




char* insertrnstring(char** buffer, const char* src, size_t pos, size_t len)
{
	size_t src_len;
	
	if(!src)
	{
		ss_errno = SS_NULL_BUFFER;
		return NULL;
	}

	if(!len)
	{
		ss_errno = SS_NO_ERROR;
		return *buffer;
	}

	src_len = strlen(src);

	if(len > src_len) len = src_len;

	src += src_len - len;

	return insertstring(buffer, src, pos);
}


char* replacestring(char** buffer, const char* haystack, const char* needle)
{
	/* wrapper of replacenstring 1*/
	return replacenstring(buffer, haystack, needle, needle ? strlen(needle) : 0);
}



char* replacerstring(char** buffer, const char* haystack, const char* needle)
{
	return replacernstring(buffer, haystack, needle, needle ? strlen(needle) : 0);
}




char* replacenstring(char** buffer, const char* haystack, const char* needle, size_t len) /* 1 */
{
	return __replacestr(buffer, haystack, needle, len, 1);
}




char* replacenlaststring(char** buffer, const char* haystack, const char* needle, size_t len)
{
	return __replacelaststr(buffer, haystack, needle, len, 1);
}




char* replacernstring(char** buffer, const char* haystack, const char* needle, size_t len)
{
	return __replacestr(buffer, haystack, needle, len, 0);
}




char* replacernlaststring(char** buffer, const char* haystack, const char* needle, size_t len)
{
	return __replacelaststr(buffer, haystack, needle, len, 0);
}



int emptystring(const char* src)
{
	size_t len, i;
	if(!src)
		return -1;

	if(!*src)
		return 1;

	len = strlen(src);

	for(i=0; i<len; ++i)
	{
		if(!isspace(src[i]))
			return 0;
	}

	return 1;
}
